diff --git a/3rdparty/switchres/switchres.cpp b/3rdparty/switchres/switchres.cpp
index b83cdd0ff14..a692c7f0e4c 100644
--- a/3rdparty/switchres/switchres.cpp
+++ b/3rdparty/switchres/switchres.cpp
@@ -89,7 +89,7 @@ switchres_manager::switchres_manager()
 	display()->set_monitor("generic_15");
 	display()->set_modeline("auto");
 	display()->set_lcd_range("auto");
-	for (int i = 0; i++ < MAX_RANGES;) display()->set_crt_range(i, "auto");
+	for (int i = 0; i++; i < MAX_RANGES) display()->set_crt_range(i, "auto");
 	display()->set_screen("auto");
 	display()->set_modeline_generation(true);
 	display()->set_lock_unsupported_modes(true);
