diff --git a/src/osd/sdl/sdlmain.cpp b/src/osd/sdl/sdlmain.cpp
index 3c57faeddbb..e2d14cd435c 100644
--- a/src/osd/sdl/sdlmain.cpp
+++ b/src/osd/sdl/sdlmain.cpp
@@ -448,8 +448,8 @@ void sdl_osd_interface::init(running_machine &machine)
 		{
 #if defined(__linux__)
 			// bgfx does not work with wayland
-			osd_printf_verbose("Setting SDL videodriver '%s' ...\n", "x11");
-			osd_setenv(SDLENV_VIDEODRIVER, "x11", 1);
+			osd_printf_verbose("Setting SDL videodriver '%s' ...\n", "x11,kmsdrm");
+			osd_setenv(SDLENV_VIDEODRIVER, "x11,kmsdrm", 1);
 #endif
 		}
 	}
