diff --git a/configure.ac b/configure.ac
index d0f2390..c2bbe45 100755
--- a/configure.ac
+++ b/configure.ac
@@ -132,7 +132,7 @@ dnl * Needed Libraries *
 dnl ********************
 AC_MSG_CHECKING([dependencies])
 
-PKG_CHECK_MODULES([LIBUSB], [libusb-1.0 >= 1.0.~],
+PKG_CHECK_MODULES([LIBUSB], [libusb-1.0 >= 1.0],
 	echo -e "\tlibusb 1.0 ver `pkg-config --short-errors --modversion libusb-1.0` found",
 	AC_MSG_ERROR([libusb-1.0 development library not found.])
 )
