diff --git a/repos/extra-x86_64/PKGBUILD b/repos/extra-x86_64/PKGBUILD
index fafb738b903..200acfa24f7 100644
--- a/repos/extra-x86_64/PKGBUILD
+++ b/repos/extra-x86_64/PKGBUILD
@@ -9,7 +9,7 @@ pkgdesc="2D graphics library with support for multiple output devices"
 url="https://cairographics.org/"
 arch=(x86_64)
 license=(LGPL MPL)
-depends=(lzo zlib libpng fontconfig freetype2 libx11 libxext libxrender libxcb
+depends=(lzo zlib libpng fontconfig freetype2 systemd libdrm libgl
          glib2 pixman)
 makedepends=(valgrind git meson)
 _commit=156cd3eaaebfd8635517c2baf61fcf3627ff7ec2  # tags/1.17.4^0
@@ -43,7 +43,10 @@ build() {
   arch-meson cairo build \
     -D spectre=disabled \
     -D tee=enabled \
-    -D tests=disabled
+    -D tests=disabled \
+    -D gl-backend=gl \
+    -D xcb=disabled \
+    -D xlib=disabled
   meson compile -C build
 }
 
