diff --git a/linux/trunk/config b/linux/trunk/config
index f9f36d1da9e..f1ae1fe63cf 100644
--- a/linux/trunk/config
+++ b/linux/trunk/config
@@ -5983,7 +5983,7 @@ CONFIG_DRM=m
 CONFIG_DRM_MIPI_DSI=y
 CONFIG_DRM_DP_AUX_CHARDEV=y
 # CONFIG_DRM_DEBUG_SELFTEST is not set
-CONFIG_DRM_KMS_HELPER=m
+CONFIG_DRM_KMS_HELPER=y
 CONFIG_DRM_KMS_FB_HELPER=y
 CONFIG_DRM_FBDEV_EMULATION=y
 CONFIG_DRM_FBDEV_OVERALLOC=100
