diff --git a/repos/extra-x86_64/PKGBUILD b/repos/extra-x86_64/PKGBUILD
index cf1cd64e018..686b5792c8f 100644
--- a/repos/extra-x86_64/PKGBUILD
+++ b/repos/extra-x86_64/PKGBUILD
@@ -16,11 +16,21 @@ optdepends=('alsa-lib: ALSA audio driver'
             'jack: JACK audio driver'
             'pipewire: PipeWire audio driver'
             'libdecor: Wayland client decorations')
-source=("https://www.libsdl.org/release/SDL2-${pkgver}.tar.gz"{,.sig})
+source=("https://www.libsdl.org/release/SDL2-${pkgver}.tar.gz"{,.sig}
+        'force_modes_refresh.patch'
+        'SDL_x11modes.c_2.0.18_to_2.0.14.patch')
 sha512sums=('8688b855ce15d574fb0726dd2a44109bb7b34d81c652c51025a32bf7eb7015ceb685834cf9fc3f97a2f5f5a3203f548ce6845420fafae7e5dc6bb9d11ce1740d'
+            'SKIP'
+            'SKIP'
             'SKIP')
 validpgpkeys=('1528635D8053A57F77D1E08630A59377A7763BE6') # Sam Lantinga
 
+prepare() {
+  cd SDL2-${pkgver}
+  patch -Np1 -i "$srcdir"/force_modes_refresh.patch
+  patch -Np1 -i "$srcdir"/SDL_x11modes.c_2.0.18_to_2.0.14.patch
+}
+
 build() {
   cd SDL2-${pkgver}
   cmake \
