diff --git a/repos/extra-x86_64/PKGBUILD b/repos/extra-x86_64/PKGBUILD
index 4f799db39f4..1065c74796f 100644
--- a/repos/extra-x86_64/PKGBUILD
+++ b/repos/extra-x86_64/PKGBUILD
@@ -15,11 +15,24 @@ optdepends=('alsa-lib: ALSA audio driver'
             'jack: JACK audio driver'
             'pipewire: PipeWire audio driver'
             'libdecor: Wayland client decorations')
-source=("https://www.libsdl.org/release/SDL2-${pkgver}.tar.gz"{,.sig})
+source=("https://www.libsdl.org/release/SDL2-${pkgver}.tar.gz"{,.sig}
+        "force_modes_refresh.patch"
+        "sdl_fix_udev_bug.patch"
+        "sdl_restore_crtc_at_DestroyWindow.patch")
 sha512sums=('ca4b690433cd4d9d73b797da98666317128e7e817ab60e874a49d94791ea41e8a6b4fc43649593120daa0702190c0f8a6ed326c908d87375c8da9f369d994f6a'
+            'SKIP'
+            'SKIP'
+            'SKIP'
             'SKIP')
 validpgpkeys=('1528635D8053A57F77D1E08630A59377A7763BE6') # Sam Lantinga
 
+prepare() {
+  cd SDL2-${pkgver}
+  patch -Np1 -i "$srcdir"/force_modes_refresh.patch
+  patch -Np1 -i "$srcdir"/sdl_fix_udev_bug.patch
+  patch -Np1 -i "$srcdir"/sdl_restore_crtc_at_DestroyWindow.patch
+}
+
 build() {
   CFLAGS+=" -ffat-lto-objects"
   cmake -S SDL2-${pkgver} -B build -G Ninja \
