diff --git a/repos/extra-x86_64/PKGBUILD b/repos/extra-x86_64/PKGBUILD
index 4f799db39f4..1065c74796f 100644
--- a/repos/extra-x86_64/PKGBUILD
+++ b/repos/extra-x86_64/PKGBUILD
@@ -15,11 +15,23 @@ optdepends=('alsa-lib: ALSA audio driver'
             'jack: JACK audio driver'
             'pipewire: PipeWire audio driver'
             'libdecor: Wayland client decorations')
-source=("https://github.com/libsdl-org/SDL/releases/download/release-${pkgver}/SDL2-${pkgver}.tar.gz"{,.sig})
+source=("https://github.com/libsdl-org/SDL/releases/download/release-${pkgver}/SDL2-${pkgver}.tar.gz"{,.sig}
+        "force_modes_refresh.patch"
+        "sdl_fix_udev_bug.patch"
+        "sdl_restore_crtc_at_DestroyWindow.patch")
 sha512sums=('26c10fe004e47a498878ff2aa80bab7f57798d6b0f950b085409818d4c758010be7b49f5e0f43dd0afe2b9d18f94ad52d1e212a1d55505ce6d7fcc371fdda7d8'
+            'SKIP'
+            'SKIP'
+            'SKIP'
             'SKIP')
 validpgpkeys=('1528635D8053A57F77D1E08630A59377A7763BE6') # Sam Lantinga
 
+prepare() {
+  cd SDL2-${pkgver}
+  patch -Np1 -i "$srcdir"/force_modes_refresh.patch
+  patch -Np1 -i "$srcdir"/sdl_restore_crtc_at_DestroyWindow.patch
+}
+
 build() {
   CFLAGS+=" -ffat-lto-objects"
   cmake -S SDL2-${pkgver} -B build -G Ninja \
