diff --git a/mame/repos/community-x86_64/PKGBUILD b/mame/repos/community-x86_64/PKGBUILD
index 3832088758c..a616e3d38d9 100644
--- a/mame/repos/community-x86_64/PKGBUILD
+++ b/mame/repos/community-x86_64/PKGBUILD
@@ -3,51 +3,48 @@
 # Contributor: robb_force <robb_force@holybuffalo.net>
 # Contributor: JJDaNiMoTh <jjdanimoth@gmail.com>
 
-pkgname=mame
+pkgname=groovymame
 pkgver=0.219
 pkgrel=1
 pkgdesc="A port of the popular Multiple Arcade Machine Emulator using SDL with OpenGL support."
+_mamever=${pkgver/./}
+_switchresver=017p
 url="https://mamedev.org/"
 license=(GPL2)
 arch=(x86_64)
-depends=(sdl2_ttf qt5-base lua libutf8proc pugixml portmidi portaudio flac)
+depends=(sdl2_ttf qt5-base lua libutf8proc pugixml portmidi portaudio flac libxrandr)
 makedepends=(nasm python asio rapidjson glm libxinerama)
 conflicts=(sdlmame)
 replaces=(sdlmame)
-source=("https://github.com/mamedev/mame/archive/mame${pkgver/./}.tar.gz"
+source=( "https://github.com/antonioginer/mame/archive/groovymame${_mamever}.tar.gz"
+         "https://github.com/antonioginer/GroovyMAME/releases/download/gm0219/groovymame64_${_mamever}.${_switchresver}_linux.tar.bz2"
          mame.sh mame.desktop mame.svg)
-sha256sums=('9513c046033e451d86adfdd297da8a2ad9b6642a2f3ef669e1443d0acc2c2e0b'
+sha256sums=(SKIP
+            SKIP
             'ee1c59bafc5e5441e99fa4c58108a3e18048e60672f34de865c8a5a976094dba'
             '6beb883c8efed5b7466d43d0658b47c3e4a9928b5d0245ed56446b230e28306b'
             '17c442c933d764175e4ce1de50a80c0c2ddd5d733caf09c3cd5e6ba697ac43f4')
 
 
-prepare() {
-  cd mame-mame${pkgver/./}
-
-# Use system libraries
-  sed -e 's|\# USE_SYSTEM_LIB|USE_SYSTEM_LIB|g' -i makefile
-# except for asio
-  sed -e 's|USE_SYSTEM_LIB_ASIO|\# USE_SYSTEM_LIB_ASIO|g' -i makefile
-}
-
 build() {
-  cd mame-mame${pkgver/./}
-  make -j1 \
+  cd GroovyMAME-${pkgname}${_mamever}
+  make \
     NOWERROR=1 \
     OPTIMIZE=2 \
     TOOLS=1 \
+    PTR64=1 NOASM=0 TARGET=mame SUBTARGET=tiny \
     ARCHOPTS=-flifetime-dse=1
 }
 
 package() {
-  cd mame-mame${pkgver/./}
+  cd GroovyMAME-${pkgname}${_mamever}
 
   # Install the mame script
-  install -Dm755 "$srcdir"/$pkgname.sh "$pkgdir"/usr/bin/$pkgname
+  install -Dm755 "$srcdir"/mame.sh "$pkgdir"/usr/bin/mame
 
   # Install the binaries
-  install -Dm755 mame64 "$pkgdir"/usr/lib/mame/mame
+  install -Dm755 "$srcdir"/groovymame "$pkgdir"/usr/lib/mame/${pkgname}
+  ln -s ./groovymame "$pkgdir"/usr/lib/mame/mame
   for _i in castool chdman imgtool jedutil nltool nlwav pngcmp regrep romcmp src2html \
             split srcclean ldverify ldresample; do
     install -m755 $_i -t "$pkgdir"/usr/lib/mame
@@ -55,8 +52,10 @@ package() {
   done
 
   # Install the extra bits
-  install -Dm644 src/osd/modules/opengl/shader/glsl*.*h -t "$pkgdir"/usr/lib/$pkgname/shader/
-  cp -ar {artwork,bgfx,plugins,language,ctrlr,keymaps,hash} "$pkgdir"/usr/lib/$pkgname/
+  install -Dm644 src/osd/modules/opengl/shader/glsl*.*h -t "$pkgdir"/usr/lib/mame/shader/
+  cp -ar {artwork,bgfx,plugins,language,ctrlr,keymaps,hash} "$pkgdir"/usr/lib/mame/
+  mkdir -p "$pkgdir"/usr/share/mame
+  "$pkgdir"/usr/lib/mame/groovymame -listxml > "$pkgdir"/usr/share/mame/mame.dat
 
   # Include the license
   install -Dm644 docs/LICENSE "$pkgdir"/usr/share/licenses/$pkgname/LICENSE
