diff --git a/linux/repos/core-x86_64/config b/linux/repos/core-x86_64/config
index 14f9a5869e7..60527113d43 100644
--- a/linux/repos/core-x86_64/config
+++ b/linux/repos/core-x86_64/config
@@ -23,7 +23,7 @@ CONFIG_INIT_ENV_ARG_LIMIT=32
 # CONFIG_COMPILE_TEST is not set
 # CONFIG_HEADER_TEST is not set
 CONFIG_LOCALVERSION=""
-CONFIG_LOCALVERSION_AUTO=y
+CONFIG_LOCALVERSION_AUTO=n
 CONFIG_BUILD_SALT=""
 CONFIG_HAVE_KERNEL_GZIP=y
 CONFIG_HAVE_KERNEL_BZIP2=y
