diff --git a/repos/extra-x86_64/PKGBUILD b/repos/extra-x86_64/PKGBUILD
index b2e5031ab2a..cda7ecf01d6 100644
--- a/repos/extra-x86_64/PKGBUILD
+++ b/repos/extra-x86_64/PKGBUILD
@@ -7,7 +7,7 @@ pkgdesc="The GL Vendor-Neutral Dispatch library"
 arch=('x86_64')
 url="https://github.com/NVIDIA/libglvnd"
 license=('custom:BSD-like')
-makedepends=('libxext' 'libx11' 'xorgproto' 'python' 'meson')
+makedepends=('python' 'meson')
 provides=('libgl' 'libegl' 'libgles')
 source=("https://gitlab.freedesktop.org/glvnd/libglvnd/-/archive/v$pkgver/libglvnd-v$pkgver.tar.gz"
         LICENSE)
@@ -16,7 +16,9 @@ sha512sums=('ca741bdd7ac2fba919f1f85e685e92396bc1cdc436e77bca921d87738bb273a6d48
 
 build() {
   arch-meson $pkgname-v$pkgver build \
-    -D gles1=false
+    -D gles1=false \
+    -D x11=disabled \
+    -D glx=disabled
 
   ninja -C build
 }
