diff --git a/PKGBUILD b/PKGBUILD
index 45c6c46..c13ee6d 100644
--- a/PKGBUILD
+++ b/PKGBUILD
@@ -15,12 +15,28 @@ conflicts=('attractmode-git')
 source=("https://github.com/mickelson/${pkgname}/archive/v${pkgver}.tar.gz"
 	"https://github.com/mickelson/${pkgname}/releases/download/v1.6.2/ATTRACT.MODE.intro.16-9.v6.1080p.mp4"
 	"https://github.com/mickelson/${pkgname}/releases/download/v1.6.2/ATTRACT.MODE.intro.4-3.v6.1080p.mp4"
-	attract.desktop)
+	attract.desktop
+	0001-Fixed-frame-delay-on-nested-surfaces.patch
+	0002-Fixed-frame-delay-on-nested-surfaces-updated.patch
+	0003-Added-Power-Saving-option-to-general-settings.patch
+	)
 sha256sums=('4c17641ce0a0b9ff77da5e75f0cb26cc48e4c6295ef57d86a37db18b2bd9fbe9'
             '0a21286df3bd51edd24f6eb21a246f8a6576020d34ce3f5cfe93304dbbadc23b'
             '5750cf26864a86d7625f52223ac1a4050ff4475a016bb4f94fc506b7558dbaf9'
             'd5e94d30bf6329bef87f58aad31ae9ff4632550a8576c243e41f220262c36cf6')
 
+prepare () {
+  cd "${pkgname}-${pkgver}"
+  local src
+  for src in "${source[@]}"; do
+    src="${src%%::*}"
+    src="${src##*/}"
+    [[ $src = *.patch ]] || continue
+    echo "Applying patch $src..."
+    patch -Np1 < "../$src"
+  done
+}
+
 build() {
 	cd "${pkgname}-${pkgver}"
 	VERSION="$pkgver" make prefix=/usr
