diff --git a/PKGBUILD b/PKGBUILD
index fd4d019..96d80d2 100644
--- a/PKGBUILD
+++ b/PKGBUILD
@@ -6,7 +6,7 @@ pkgdesc="A cross-platform frontend designed for MAME cabinets/game centers/etc.
 arch=('i686' 'x86_64')
 url="http://www.retrofe.com"
 license=('GPL3')
-depends=('glib2' 'zlib' 'gstreamer0.10-good' 'sdl2_image' 'sdl2_mixer' 'sdl2_ttf')
+depends=('glib2' 'zlib' 'gstreamer' 'gst-plugins-base' 'sdl2_image' 'sdl2_mixer' 'sdl2_ttf')
 makedepends=('cmake' 'mercurial' 'python2')
 source=('hg+https://phulshof@bitbucket.org/phulshof/retrofe#branch=RetroFE-Release-0.9.0'
 	$pkgname.desktop
