diff --git a/PKGBUILD b/PKGBUILD
index cdb3003..e026935 100644
--- a/PKGBUILD
+++ b/PKGBUILD
@@ -16,6 +16,15 @@ install=${pkgname}.install
 source=("https://github.com/panzi/$pkgname/archive/v$pkgver.tar.gz")
 md5sums=('1e8084db56dacdb863ca5e5e40f0d640')
 
+prepare() {
+  local src
+  pwd
+  for src in ../*.patch ; do
+    msg2 "Applying patch $src..."
+    patch -d ${pkgname}-${pkgver} -Np1 < "$src"
+  done
+}
+
 package() {
   cd "${srcdir}/${pkgname}-${pkgver}/"
   mkdir -p build
